### PR TITLE
[Tests] Diverse Whisper fixes

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

There's a lot of pending failing tests with Whisper. This PR addresses some issues:

1. #31683 and #31770 mentioned a out-of-range word level timestamps. This happens because `decoder_inputs_ids` were once `forced_input_ids`. This had an impact on the `beam_indices`. 

`beam_indices` has a length of `decoder_input_ids + potentially_generated_ids` but doesn't take into account `decoder_input_ids` when keeping track of the indices. In other words `beam_indices[0]` is really the beam indice of the first generated token, instead of `decoder_input_ids[0]`.

2. The Flash-Attention 2 attention mask was causing an issue

3. The remaining work is done on the modeling tests. Note that some of these tests were failing because of straightforward reasons - e.g the output was a dict -  and are actually still failing, but their reasons for failing are not straightforward anymore. Debugging will be easier though.

**Note:** With #33450 and this, we're down from 29 failing tests to 17

### CODE REVIEW

**Comment by LysandreJik** [comment]
LMK when you want me to review @ylacombe (please fix the tests and the conflicts first :blush:)

**Comment by ylacombe** [comment]
Hey @LysandreJik, could you review now ? It seems that the failing check is unrelated to Whisper !

**Review by LysandreJik — APPROVED** [review]
I'm not totally up to date with the changes in `generation_whisper` so I'll trust you with them; if you want a double check feel free to cc @ArthurZucker but if you're confident you can ahead and merge.

**Comment by ylacombe** [comment]
cc @ArthurZucker  for a quick look !

**Review by ArthurZucker — COMMENTED** [review]
LGTM, let's just run slow before we merge? 🤗

**Comment by ylacombe** [comment]
cc @ydshieh , there's still this error when doing slow tests: 
 ```
File "/usr/lib/python3.8/runpy.py", line 194 in _run_module_as_main
Bus error (core dumped)
```

**Comment by ydshieh** [comment]
hi @ylacombe let me merge #33849 and try to rebase this PR on top of main and let's see.

**Comment by ylacombe** [comment]
It's working thanks !

**Comment by ylacombe** [comment]
The list of failing tests are as expected, merging !

**Comment by leng-yue** [comment]
Great work! I noticed there's a small issue in this PR, especially when `prompt_ids` is specified—the model always returns a timestamp of 0.0.

To fix this quickly, you can update https://github.com/huggingface/transformers/blob/38f9f10dd9240619ea17fb6c7acb51b3bc592232/src/transformers/models/whisper/generation_whisper.py#L191 to:
```python
cross_attentions.append(torch.cat([x[i] for x in generate_outputs.cross_attentions], dim=2)[:, :, num_input_ids-3:, :])
```

**Comment by ylacombe** [comment]
Hey @leng-yue, thanks for your message!
While there's maybe something wrong with how cross-attentions weights are computed, I'm not sure that your proposed solution works. 
Could you provide a code snippet and maybe an explanation of the proposed solution if you have time?
Thanks!

**Comment by leng-yue** [comment]
Basically `num_input_ids` is `prefix_tokens` (timestamps, language, etc) + `prompt_ids`, therefore, we want to strip the`prompt_ids`, which is `num_input_ids` - 3.
